### DISCUSSION TITLE
File storage component on loopback4

### DISCUSSION THREAD

**Discussion by emazzu** [loopbackio/loopback-next#7283] [Q&A] [answered]
Hi people, 
I don't find documentation for file system storage, I m looking for loopback-next/examples/file-transfer/, but i dont know, how to create the datasources. 

could somebody help my ?? please. !!!

Becouse the connector to database is working OK, 

Regards
Eduardo

**Comment by marioestradarosa** [comment, marked as answer]
@emazzu , the example shows how to provide upload/download. **Multer** is used as the storage, but keep in mind that this storage is a folder in your file system. At the end of `application.ts` file you can find the multer options.  However, the dataSource as per a database connection is up to you, this example shows how to manage the appropriate upload/download of a file and served from a loopback 4 server.  

additionally you can have a database table that associates the file name to a record but that s another standard step in loopback 4 using the concept of datasource/models/repository. Keep in mind that these 3 artifacts don't necessarily will have to have a controller, in this case you can use them in the upload/download controller also.

**Reply by emazzu** [reply]
@marioestradarosa 

This is de link, you did reference ??
@loopback/example-file-transfer
https://github.com/strongloop/loopback-next/tree/master/examples/file-transfer

**Reply by marioestradarosa** [reply]
@emazzu , that-s correct, you can run `lb4 example file-transfer` and get that working example on your computer. The example only shows how to properly upload/download a file using the framework. The files are stored in a physical directory on your computer (see multer options). In a real scenario you usually link this file somehow to a record in a table (ie: users uploading/downloading files).  

Your userfiles table might contain userid and file_name as a reference only, but in reality the file resides on disk.

**Reply by emazzu** [reply]
@marioestradarosa 
Yes, in this case I have the volume mount on linux server, where the lb4 will be deployed and it 's managed as a local file server. 
The example is good for me !!!

And I have other endpoint with some parameters for access to documents.

example: 
![image](https://user-images.githubusercontent.com/6100546/113728452-8099f400-96cc-11eb-8a7c-dc50a90da0e2.png)

**Comment by emazzu** [comment]
@marioestradarosa 
I send endpoint for some other user who needs it

This controller is connected with a datasource, because I make a query, and it returns data where is the file.
method @get('/documents')
method @get('/documents/lixian-files')

The build the path, is for windows, I think for linux, I shoud change samething.

```
import {inject} from '@loopback/core';
import {
  Filter,
  repository
} from '@loopback/repository';
import {
  get,
  getModelSchemaRef,
  HttpErrors,
  oas,
  param,
  Response,
  response, RestBindings
} from '@loopback/rest';
import {Documents} from '../models/lixian-documents.model';
import {DocumentsRepository} from '../repositories/lixian-documents.repository';


export class DocumentsController {
  constructor(
    @repository(DocumentsRepository)
    public lixianDocumentsRepository: DocumentsRepository,
  ) { }

  @get('/documents')
  @response(200, {
    description: 'Array of LixianDocuments model instances',
    content: {
      'application/json': {
        schema: {
          type: 'array',
          items: getModelSchemaRef(Documents, {includeRelations: true}),
        },
      },
    },
  })
  async find(
    @param.filter(Documents) filter?: Filter<Documents>,
  ): Promise<Documents[]> {
    return this.lixianDocumentsRepository.find(filter);
  }


  // metodo personalizado
  //
  @get('documents/lixian-files/{serverPath}/{volume}/{id}/{fileName}')
  @oas.response.file()
  downloadFile(
    @param.path.string('serverPath') serverPath: string,
    @param.path.string('volume') volume: string,
    @param.path.string('id') id: string,
    @param.path.string('fileName') fileName: string,
    @inject(RestBindings.Http.RESPONSE) response: Response,
  ) {

    const SANDBOX = '\\' + 'users' + '\\' + 'xmazzuce' + '\\' + 'lb4' + '\\' + 'LIXIAN-DOCS' + '\\' + serverPath + '\\' + volume + '\\' + id;

    const file = validateFileName(SANDBOX, fileName);
    response.download(file, fileName);
    return response;
  }

}

//
// Validate
//
function validateFileName(sandBox: string, fileName: string) {
  // la linea comentada es la original, recomendada x loopback, pero la reemplace
  //const resolved = path.resolve(sandBox, fileName);
  const resolved = sandBox + '\\' + fileName;
  if (resolved.startsWith(sandBox)) return resolved;
  // The resolved file is outside sandbox
  throw new HttpErrors.BadRequest(`Invalid file name: ${fileName}`);
}
```
